features/steps/steps.py

```diff
@@ -283,4 +294,19 @@ def step_impl(context, representation_id):
                     errors.append(representation_shape_error(inst, representation_id))
     
     handle_errors(context, errors)
+
+@then('Each {entity} may be nested by only the following entities: {other_entities}')
+def step_impl(context, entity, other_entities):
+
+    allowed_entity_types = other_entities.split(', ')
+
+    errors = []
+    if getattr(context, 'applicable', True):
+        for inst in context.model.by_type(entity):
+            nested_entities = [i for rel in inst.IsNestedBy for i in rel.RelatedObjects]
+            nested_entity_types = [i.is_a() for i in nested_entities]
+            if not set(nested_entity_types) <= set((allowed_entity_types)):
+                differences = list(set(nested_entity_types) - set(allowed_entity_types))
+                errors.append(instance_structure_error(inst, [i for i in nested_entities if i.is_a() in differences], 'nested by',{}))
```

Comment:
Same. You already have default_factory=dict. I would leave it out here. Looks prettier.

But what is the use of `optional_values` then actually?

Reply:
The use of optional_values is to avoid merge conflicts between ALB001 and ALB002/ALB003 https://github.com/buildingSMART/ifc-gherkin-rules/commits/alb001_alignment-in-spatial-structure
We used different dataclasses and it was better to merge them. @JTurbakiewicz used sentences as
`A is/is not directly/indirectly contained in B` 
'is/isnot' and 'directly/indirectly' play a role in ALB001, but not in the current PR.